src/Utils.php

```diff
@@ -307,9 +307,10 @@ public static function streamFor($resource = '', array $options = [])
 
         switch (gettype($resource)) {
             case 'resource':
-                if (\stream_get_meta_data($resource)['uri'] === 'php://input') {
+                if (in_array(\stream_get_meta_data($resource)['uri'], ['php://input', 'php://memory'])) {
                     $stream = self::tryFopen('php://temp', 'w+');
                     fwrite($stream, stream_get_contents($resource));
+                    fseek($stream, 0);
```

Comment:
I still dont think this is the proper patch. I think it "just solves the problem" without solving the root cause.

Reply:
```
$f = fopen('php://input', 'r');
var_dump(ftell($f));
$stream = \GuzzleHttp\Psr7\Utils::streamFor($f);
$stream->tell();
var_dump($stream->tell());
```
returns
```
int 0
int 3
```

In my opinion moving the pointer at time of creation the stream is a bug.

Reply:
No.. the bug is that `StreamInterface::__toString()` is not returning the full string. This is the fact we should focus on. As you see from the implementation, we always do `Stream::seek(0)` in `__toString()`, however, that gives no effect. Im super curious why and how to fix that.

Reply:
As you wish.

I think that moving the pointer is not obvious behaviour and it should be fixed.

when you create a stream from file path (like `Utils::streamFor(__FILE__)`) the pointer is at 0
when you create a stream from `'php://input'` resource the pointer is at last byte. 

Why?

Reply:
Btw for scalar type it's clear and understandable:
https://github.com/guzzle/psr7/blob/master/src/Utils.php#L286

When you create the stream from file path you expect to have pointer at 0. Even for sake of consistency I would move pointer of `'php://input'` stream to 0

Reply:
> Even for sake of consistency I would move pointer of 'php://input' stream to 0

No, I don't think we can do that, certainly not in 1.x. Would be breaking, and we also don't know if it was at zero to begin with.

Reply:
But introduction of this new behaviour was not a breaking change, right ?

That does not matter if it was not at zero before. Php input resource read from the position of current pointer. But for the new stream created it's 0 byte. Please see the unit test.